compiler/plugins/target/AMD-AIE/aie/AMDAIECreatePathFindFlows.cpp

```diff
@@ -13,671 +14,566 @@
-                                                 getTile(builder, col, row));
-  SwitchboxOp::ensureTerminator(switchboxOp.getConnections(), builder,
+SwitchboxOp getOrCreateSwitchbox(OpBuilder &builder, DeviceOp &device, int col,
+                                 int row) {
+  auto tile = getOrCreateTile(builder, device, col, row);
+  for (auto i : tile.getResult().getUsers()) {
+    if (llvm::isa<SwitchboxOp>(*i)) return llvm::cast<SwitchboxOp>(*i);
+  }
+  OpBuilder::InsertionGuard g(builder);
+  auto sbOp = builder.create<SwitchboxOp>(builder.getUnknownLoc(), tile);
+  SwitchboxOp::ensureTerminator(sbOp.getConnections(), builder,
                                 builder.getUnknownLoc());
-  coordToSwitchbox[{col, row}] = switchboxOp;
-  maxCol = std::max(maxCol, col);
-  maxRow = std::max(maxRow, row);
-  return switchboxOp;
+  return sbOp;
 }
 
-ShimMuxOp DynamicTileAnalysis::getShimMux(OpBuilder &builder, int col) {
-  assert(col >= 0 && "expected col >= 0");
-  int row = 0;
-  if (coordToShimMux.count({col, row})) return coordToShimMux[{col, row}];
-
-  assert(getTile(builder, col, row).isShimNOCTile() &&
-         "expected tile is ShimNOC");
-  auto switchboxOp = builder.create<ShimMuxOp>(builder.getUnknownLoc(),
-                                               getTile(builder, col, row));
-  SwitchboxOp::ensureTerminator(switchboxOp.getConnections(), builder,
-                                builder.getUnknownLoc());
-  coordToShimMux[{col, row}] = switchboxOp;
-  maxCol = std::max(maxCol, col);
-  maxRow = std::max(maxRow, row);
-  return switchboxOp;
+ShimMuxOp getOrCreateShimMux(OpBuilder &builder, DeviceOp &device, int col) {
+  auto tile = getOrCreateTile(builder, device, col, /*row*/ 0);
+  for (auto i : tile.getResult().getUsers()) {
+    if (auto shim = llvm::dyn_cast<ShimMuxOp>(*i)) return shim;
+  }
+  OpBuilder::InsertionGuard g(builder);
+  auto shmuxOp = builder.create<ShimMuxOp>(builder.getUnknownLoc(), tile);
+  ShimMuxOp::ensureTerminator(shmuxOp.getConnections(), builder,
+                              builder.getUnknownLoc());
+  return shmuxOp;
 }
 
-// allocates channels between switchboxes ( but does not assign them)
-// instantiates shim-muxes AND allocates channels ( no need to rip these up in )
 struct ConvertFlowsToInterconnect : OpConversionPattern<FlowOp> {
   using OpConversionPattern::OpConversionPattern;
-  DynamicTileAnalysis &analyzer;
-  ConvertFlowsToInterconnect(MLIRContext *context, DynamicTileAnalysis &a,
-                             PatternBenefit benefit = 1)
-      : OpConversionPattern(context, benefit), analyzer(a) {}
+  const std::map<PathEndPoint, SwitchSettings> flowSolutions;
+  std::set<PathEndPoint> processedFlows;
+
+  ConvertFlowsToInterconnect(
+      MLIRContext *context,
+      const std::map<PathEndPoint, SwitchSettings> &flowSolutions,
+      std::set<PathEndPoint> &processedFlows, PatternBenefit benefit = 1)
```

Comment:
```suggestion
      const std::set<PathEndPoint> &processedFlows, PatternBenefit benefit = 1)
```

Reply:
Nope - it's mutated in here (don't ask - I tried to get it out and it's too cumbersome so I left it go for the moment).